### PR TITLE
WIP: Allow custom thresholds in annotate_flat()

### LINKED ISSUE #10137: Allow custom thresholds in annotate_flat()
From https://mne.discourse.group/t/annotating-raw-data-based-on-amplitude-thresholds/4143/3

> I actually think that we should enhance `annotate_flat` such that one can specify a flatness threshold, similarly to what we offer when creating epochs. Currently we only [check](https://github.com/mne-tools/mne-python/blob/c3ff05c05c7bfcf765249f27a517803f27f48b4e/mne/preprocessing/flat.py#L74) whether the data is actually **absolutely** flat, i.e., the difference between samples is zero. It should be easy to allow for an absolute non-zero threshold here instead.

cc @sappelhoff @mscheltienne

### PR DESCRIPTION
**Pull request by mscheltienne**
Closes #10137 

@hoechenberger WDYT of a dictionary per channel type? 
Name OK? Or would you prefer `flat` to match `Epochs`?

Waiting for CIs to confirm I didn't break anything; TODO:
- [ ] add entry to changelog
- [ ] add tests

### CODE REVIEW

**Comment by sappelhoff** [comment]
Too bad this is called annotate_flat. Generalizing this to annotate both extremely high AND extremely low ptp segments would be useful.

**Comment by mscheltienne** [comment]
I was thinking about it.. I guess `annotate_flat` is not that used, no?

**Comment by hoechenberger** [comment]
> EDIT: Could it maybe even be worth deprecating `annotate_flat` and replacing it with a more general function that can do the above? @mscheltienne @hoechenberger?

We could have `annotate_ptp` that does both, and `annotate_flat` could internally just call this function. Would allow for a long deprecation cycle if we want to go that route

File: mne/preprocessing/flat.py [hunk 1]
```diff
@@ -33,6 +33,15 @@ def annotate_flat(raw, bad_percent=5., min_duration=0.005, picks=None,
         time samples with exactly the same value are not totally uncommon.
         Defaults to 0.005 (5 ms).
     %(picks_good_data)s
+    flatness : dict | float
```
Comment (agramfort) [inline mne/preprocessing/flat.py:36]: this param is called flat in https://mne.tools/stable/generated/mne.Epochs.html?highlight=epochs#mne.Epochs

shall we be consistent here and call it flat here too?

File: mne/preprocessing/flat.py [hunk 2]
```diff
@@ -33,6 +33,15 @@ def annotate_flat(raw, bad_percent=5., min_duration=0.005, picks=None,
...
+        Reject segments based on **minimum** peak-to-peak signal amplitude
+        (PTP). Valid **keys** can be any channel type present in the object.
+        The **values** are floats that set the minimum acceptable PTP. If the
+        PTP is smaller than this threshold, the segment will be annotated.
+        If float, the minimum acceptable PTP is applied to all channels.
+        Default to ``0.`` to annotate completely flat segments.
```
Comment (agramfort) [inline mne/preprocessing/flat.py:42]: this param is called flat in https://mne.tools/stable/generated/mne.Epochs.html?highlight=epochs#mne.Epochs

shall we be consistent here and call it flat here too?

**Comment by drammock** [comment]
> We could have `annotate_ptp` or `annotate_amplitude` that does both, and `annotate_flat` could internally just call this function. Would allow for a long deprecation cycle if we want to go that route

I like this plan. I like names `annotate_amplitude` or `annotate_threshold` maybe? Params could be `peak` and `flat`, each taking a dict of channel types?

**Comment by mscheltienne** [comment]
Let's give the evening for anyone against this proposal to speak his mind; and if all are still in favor tomorrow; I'll open a PR to add it.

**Comment by agramfort** [comment]
> I like this plan. I like names annotate_amplitude or annotate_threshold maybe? Params could be peak and flat, each taking a dict of channel types?

ok for me

**Pull request closed without merging**